[mlir][XeGPU][Transform] Add vectorlinearize transform pass.

mshahneo opened the pull request:
Use upstream patterns to create a vectorlinearize pass needed for lowering to XeVM.
Linearizes n-D vectors to 1-D vectors.

This is needed because, `vector-to-llvm` does not linearize all the vectors.

mshahneo committed:
[XeGPU][Transform] Add vectorlinearize transform pass.

Use upstream patterns to create a vectorlinearize pass needed
for lowering to xevm.
Linearizes n-D vectors to 1-D vectors.

nbpatel reviewed (commented):
can you add a couple of tests in xegpu/transform?

mshahneo committed:
Address review comments.

Add test case.

mshahneo commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,111 @@
+    target.addIllegalOp<mlir::vector::TransposeOp>();
+    target.addLegalOp<mlir::vector::ShapeCastOp>();
+    target.addLegalOp<mlir::vector::ExtractOp>();
+    target.addLegalDialect<mlir::xegpu::XeGPUDialect>();
```
Sorry, stray-code, removed, thanks for pointing this out.

mshahneo commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,111 @@
+                if (shape.size() == 2)
+                  return mlir::SmallVector<int64_t>{1, shape[1]};
+              }
+            }
```
Done.

charithaintc commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,101 @@
+//===- XeGPUVectorLinearize.cpp - Linearizes n-D vectors to 1-D vectors
+//-------===//
```
nit: best to fit this in one line :-)

charithaintc reviewed (approved):
LGTM % comments

adam-smnk commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,101 @@
+          [](mlir::Operation *op) -> std::optional<mlir::SmallVector<int64_t>> {
+            auto extractVectorType =
+                [](mlir::Operation *op) -> mlir::VectorType {
+              if (auto loadOp = mlir::dyn_cast<mlir::vector::LoadOp>(op))
```
Shouldn't nD vector load and store op be first lowered to xegpu?

mshahneo replied:
Usually yes, but not necessarily. Depending on which upper-level software stack is used, not necessarily vector.load would be converted to XeGPU load.

adam-smnk commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,101 @@
+      (void)mlir::applyPatternsGreedily(getOperation(), std::move(patterns));
+    }
+
+    // Unroll load store from <<MxN> to M <1xN> load/stores and then linearize
```
```suggestion
    // Unroll load store from <MxN> to M <1xN> load/stores and then linearize
```

mshahneo replied:
Fixed.

adam-smnk commented on mlir/test/Dialect/XeGPU/xegpu-vector-linearize.mlir:
```diff
@@ -0,0 +1,362 @@
+//       CHECK: %[[T8:.*]] = arith.index_cast %[[T6]] : vector<16xi32> to vector<16xindex>
+//       CHECK: %[[T9:.*]] = arith.index_cast %[[T7]] : vector<16xi32> to vector<16xindex>
+gpu.module @add_kernel_f32 attributes {spirv.target_env = #spirv.target_env<#spirv.vce<v1.0, [Addresses, Bfloat16ConversionINTEL, BFloat16TypeKHR, Float16Buffer, Int64, Int16, Int8, Kernel, Linkage, Vector16, GenericPointer, Groups, Float16, Float64, AtomicFloat32AddEXT, ExpectAssumeKHR, VectorAnyINTEL, VectorComputeINTEL], [SPV_EXT_shader_atomic_float_add, SPV_KHR_bfloat16, SPV_KHR_expect_assume ...
+  gpu.func @add_kernel_f32(%arg0: memref<*xf32>, %arg1: memref<*xf32>, %arg2: memref<*xf32>) kernel attributes {VectorComputeFunctionINTEL, known_block_size = array<i32: 1, 32, 1>, known_grid_size = array<i32: 1, 1, 1>, spirv.entry_point_abi = #spirv.entry_point_abi<>} {
```
I have no clue what's really validated here and the above checks only validate a part of the kernel.

I'd suggest making smaller cases that focus on some particular aspects.
In case larger examples are necessary, please add a comment above highlighting what are the expectations.

adam-smnk commented on mlir/test/Dialect/XeGPU/xegpu-vector-linearize.mlir:
```diff
@@ -0,0 +1,362 @@
+//       CHECK: %[[T7:.*]] = arith.addi %[[T5]], %[[T4]] : vector<16xi32>
+//       CHECK: %[[T8:.*]] = arith.index_cast %[[T6]] : vector<16xi32> to vector<16xindex>
+//       CHECK: %[[T9:.*]] = arith.index_cast %[[T7]] : vector<16xi32> to vector<16xindex>
+gpu.module @add_kernel_f32 attributes {spirv.target_env = #spirv.target_env<#spirv.vce<v1.0, [Addresses, Bfloat16ConversionINTEL, BFloat16TypeKHR, Float16Buffer, Int64, Int16, Int8, Kernel, Linkage, Vector16, GenericPointer, Groups, Float16, Float64, AtomicFloat32AddEXT, ExpectAssumeKHR, VectorAnyINTEL, VectorComputeINTEL], [SPV_EXT_shader_atomic_float_add, SPV_KHR_bfloat16, SPV_KHR_expect_assume ...
```
nit: line breaks (here and below) for readability

adam-smnk commented on mlir/test/Dialect/XeGPU/xegpu-vector-linearize.mlir:
```diff
@@ -0,0 +1,362 @@
+
+// -----
+// CHECK-LABEL: test_vector_transpose_16x16
+// CHECK: vector.shuffle {{.*}} [0, 16, 1, 17, 4, 20, 5, 21, 8, 24, 9, 25, 12, 28, 13, 29] : vector<16xf32>, vector<16xf32>
```
Do we really need to explicitly check shuffle lowering?
This should be covered by other tests validating respective pattern (if particular case is missing, feel free to expand other vector tests).

Perhaps simple `CHECK-COUNT-N: vector.shuffle{{.*}}: vector<16xf32>, vector<16xf32>` would be enough to just ensure unrolling happens?

adam-smnk reviewed (commented):
+1 for the pass to enable e2e lowering

I still miss relevant tests (or the current ones could use documentation) that would justify its existence and really ensure no regressions.

mshahneo committed:
Address review comments.

Update the test case to remove duplication with vector-linearize.
Add new test cases for XeGPU, vector.broadcast, vector.gather.

mshahneo commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,101 @@
+  void runOnOperation() override {
+    // vector.broadcast and vector.gather requires progressive lowering
+    {
+      mlir::RewritePatternSet patterns(&getContext());
```
Done, thanks!

mshahneo commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,101 @@
+      // Shuffle16x16 will fallback to Shuffle1D for non 16x16 sizes.
+      mlir::vector::populateVectorTransposeLoweringPatterns(
+          patterns, mlir::vector::VectorTransposeLowering::Shuffle16x16);
+      (void)mlir::applyPatternsGreedily(getOperation(), std::move(patterns));
```
Done.

mshahneo commented on mlir/test/Dialect/XeGPU/xegpu-vector-linearize.mlir:
```diff
@@ -0,0 +1,362 @@
+//       CHECK: 120, 121, 122, 123, 124, 125, 126, 127] : vector<128xf32>, vector<128xf32>
+//       CHECK: %[[RES:.*]] = vector.shape_cast %[[SHUFFLE]] : vector<64xf32> to vector<8x8xf32>
+//       CHECK: return %[[RES]] : vector<8x8xf32>
+func.func @test_extract_strided_slice_1(%arg0 : vector<8x16xf32>) -> vector<8x8xf32> {
```
Thanks for the suggestions. Fixed now, removed duplication and added some new ones.

adam-smnk commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUVectorLinearize.cpp:
```diff
@@ -0,0 +1,98 @@
+        return signalPassFailure();
+    }
+
+    // Unroll load store from <MxN> to M <1xN> load/stores and then linearize
```
It goes a bit against the promised `nD to 1D` linearization as pass fails on nD loads etc.
Maybe the pass description should include more details about the intended usage?

adam-smnk replied:
That's indeed the question, what's the intension here.

Currently pass fails on, for example, 3D `vector.load`. On one hand, it's reasonable as "naive" unrolling might not work well. Alternatively, you could enable it as a best effort try.

mshahneo committed:
Address review comments.

mshahneo commented on mlir/test/Dialect/XeGPU/xegpu-vector-linearize.mlir:
```diff
@@ -0,0 +1,285 @@
+}
+
+// -----
+// Check for vector linearization in XeGPU dialect.
```
Updated the comment.

mshahneo committed:
Address review comments.

Add vector unroll support for n-D laod/store.

adam-smnk reviewed (approved):
Looks good 👍
I'm sure it'll evolve further once we start running more e2e examples.

mshahneo committed:
Address review comments.

mshahneo commented:
> Looks good 👍 I'm sure it'll evolve further once we start running more e2e examples.

Absolutely, not just e2e, with the proposed changes in vector dialect, this would definitely evolve :)

hanhanW commented on mlir/test/Dialect/XeGPU/xegpu-vector-linearize.mlir:
```diff
@@ -0,0 +1,263 @@
+// CHECK: %17 = vector.extract %arg2[1, 0]
+// CHECK: %18 = vector.extract %arg1[1, 0]
+// CHECK: %19 = arith.addi %18, %c1
+// CHECK: %20 = scf.if %17 -> (vector<3xf32>)
```
Coming from https://github.com/llvm/llvm-project/pull/184706 that needs to update these checks. Why do you check the vars (e.g., `%17`) with numbers? This forces the future changes to update the numbers if different operations are used in some methods. Can you help fix the issue? Thanks.
